{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def validate_hmm(model):\n",
    "    \"Validates the hmm model\"\n",
    "    \n",
    "    validate = True\n",
    "    if not math.isclose(sum(model.init_probs),1):\n",
    "        validate = False\n",
    "\n",
    "    for i in range(len(model.trans_probs)):\n",
    "        count = 0\n",
    "        for j in range(len(model.trans_probs)):\n",
    "            count += model.trans_probs[i][j]\n",
    "\n",
    "        if count == 1:\n",
    "            validate = True\n",
    "\n",
    "    for i in range(len(model.emission_probs)):\n",
    "        count = 0\n",
    "        for j in range(len(model.emission_probs[0])):\n",
    "            count += model.emission_probs[i][j]\n",
    "            if (model.emission_probs[i][j] < 0.0 or model.emission_probs[i][j] > 1.0):\n",
    "                validate = False\n",
    "\n",
    "        if count == 1:\n",
    "            validate & True\n",
    "\n",
    "    return validate\n",
    "    pass\n",
    "\n",
    "def translate_indices_to_observations(indices):\n",
    "    mapping = ['a', 'c', 'g', 't']\n",
    "    return ''.join(mapping[idx] for idx in indices)\n",
    "\n",
    "def translate_path_to_indices(path):\n",
    "    return list(map(lambda x: int(x), path))\n",
    "\n",
    "def translate_indices_to_path(indices):\n",
    "    return ''.join([str(i) for i in indices])\n",
    "\n",
    "def translate_observations_to_indices(obs):\n",
    "    mapping = {'a': 0, 'c': 1, 'g': 2, 't': 3}\n",
    "    return [mapping[symbol.lower()] for symbol in obs]\n",
    "\n",
    "def read_fasta_file(filename):\n",
    "        \"\"\"\n",
    "        Reads the given FASTA file f and returns a dictionary of sequences.\n",
    "\n",
    "        Lines starting with ';' in the FASTA file are ignored.\n",
    "        \"\"\"\n",
    "        sequences_lines = {}\n",
    "        current_sequence_lines = None\n",
    "        with open(filename) as fp:\n",
    "            for line in fp:\n",
    "                line = line.strip()\n",
    "                if line.startswith(';') or not line:\n",
    "                    continue\n",
    "                if line.startswith('>'):\n",
    "                    sequence_name = line.lstrip('>')\n",
    "                    current_sequence_lines = []\n",
    "                    sequences_lines[sequence_name] = current_sequence_lines\n",
    "                else:\n",
    "                    if current_sequence_lines is not None:\n",
    "                        current_sequence_lines.append(line)\n",
    "        sequences = {}\n",
    "        for name, lines in sequences_lines.items():\n",
    "            sequences[name] = ''.join(lines)\n",
    "        return sequences\n",
    "\n",
    "class hmm:\n",
    "    def __init__(self, init_probs, trans_probs, emission_probs):\n",
    "        self.init_probs = init_probs\n",
    "        self.trans_probs = trans_probs\n",
    "        self.emission_probs = emission_probs\n",
    "    \n",
    "\n",
    "class GenomePrediction():\n",
    "    \n",
    "    def init_model():\n",
    "        \"Creation of the HMM model chosen\"\n",
    "        init_probs_7_state = [0.00, 0.00, 0.00, 1.00, 0.00, 0.00, 0.00]\n",
    "\n",
    "        trans_probs_7_state = [\n",
    "            [0.00, 0.00, 0.90, 0.10, 0.00, 0.00, 0.00],\n",
    "            [1.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00],\n",
    "            [0.00, 1.00, 0.00, 0.00, 0.00, 0.00, 0.00],\n",
    "            [0.00, 0.00, 0.05, 0.90, 0.05, 0.00, 0.00],\n",
    "            [0.00, 0.00, 0.00, 0.00, 0.00, 1.00, 0.00],\n",
    "            [0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 1.00],\n",
    "            [0.00, 0.00, 0.00, 0.10, 0.90, 0.00, 0.00],\n",
    "        ]\n",
    "\n",
    "        emission_probs_7_state = [\n",
    "            #   A     C     G     T\n",
    "            [0.30, 0.25, 0.25, 0.20],\n",
    "            [0.20, 0.35, 0.15, 0.30],\n",
    "            [0.40, 0.15, 0.20, 0.25],\n",
    "            [0.25, 0.25, 0.25, 0.25],\n",
    "            [0.20, 0.40, 0.30, 0.10],\n",
    "            [0.30, 0.20, 0.30, 0.20],\n",
    "            [0.15, 0.30, 0.20, 0.35],\n",
    "        ]\n",
    "\n",
    "        hmm_7_state = hmm(init_probs_7_state, trans_probs_7_state, emission_probs_7_state)\n",
    "        return hmm_7_state\n",
    "        pass\n",
    "    \n",
    "    def log(x):\n",
    "        if x == 0:\n",
    "            return float('-inf')\n",
    "        return math.log(x)\n",
    "\n",
    "    def compute_w_log(model, x):\n",
    "        K = len(model.init_probs)\n",
    "        N = len(x)\n",
    "\n",
    "        w = make_table(K, N)\n",
    "\n",
    "        # Base case: fill out w[i][0] for i = 0..k-1\n",
    "        # ...\n",
    "        for k in range(K):\n",
    "            w[k][0] = log(model.init_probs[k]) + log(model.emission_probs[k][x[0]])\n",
    "\n",
    "        # Inductive case: fill out w[i][j] for i = 0..k, j = 0..n-1\n",
    "        # ...\n",
    "        for n in range(1,N):\n",
    "            for j in range(0,K):\n",
    "                for i in range(0,k): #i\n",
    "                    ## take max\n",
    "                    w[j][n] = np.argmax(w[i][n-1]+log(model.trans_probs[i][j]))\n",
    "\n",
    "            w[j][n] += log(model.emission_probs[j][x[n]])\n",
    "\n",
    "        return w\n",
    "\n",
    "    def opt_path_prob_log(w):\n",
    "        return max(w[i][-1] for i in range(len(w)))\n",
    "\n",
    "    def backtrack_log(w):\n",
    "        N = len(w[0])\n",
    "        k = len(w[1])\n",
    "\n",
    "        z = [None] * N\n",
    "        z[N-1] = (argmax(w[i][-1]) for i in range(k))\n",
    "\n",
    "        for j in range(N-2,-1,-1):\n",
    "            z[j] = np.argmax(w[i][j] * log(model.trans_probs[i[z[j+1]]]) for i in range(k))\n",
    "\n",
    "        return z\n",
    "\n",
    "    def count_transitions_and_emissions(K, D, x, z):\n",
    "        \"\"\"\n",
    "        Returns a KxK matrix and a KxD matrix containing counts cf. above\n",
    "        \"\"\"\n",
    "        z_indices = translate_path_to_indices(z)\n",
    "        x_indices = translate_observations_to_indices(x)\n",
    "        trans_matrix = [ [ 0 for i in range(K) ] for j in range(K) ]\n",
    "        emi_matrix = [ [ 0 for i in range(K) ] for j in range(D) ]\n",
    "\n",
    "        for i in range(len(z_indices)-1):\n",
    "            trans_matrix[z_indices[i]][z_indices[i+1]] += 1 \n",
    "\n",
    "        for z_i,x_i in zip(z_indices,x_indices):\n",
    "            emi_matrix[z_indices[z_i]][x_indices[x_i]] += 1\n",
    "\n",
    "        return trans_matrix,emi_matrix\n",
    "        pass\n",
    "\n",
    "    def viterbi_decoding(model,x):\n",
    "        \"Wrapper function for computing w and backtracking\"\n",
    "        w = compute_w_log(model,x)\n",
    "        z = backtrack_log(w)\n",
    "        return z\n",
    "        pass\n",
    "    \n",
    "    def training_by_counting(K, D, x, z):\n",
    "        \"\"\"\n",
    "        Returns a HMM trained on x and z cf. training-by-counting.\n",
    "        \"\"\"\n",
    "        init_probs = [0.00, 0.00, 0.00, 1.00, 0.00, 0.00, 0.00]\n",
    "\n",
    "        trans, emi = count_transitions_and_emissions(K,D,x,z)\n",
    "        my_hmm = hmm(init_probs,trans,emi)\n",
    "        return hmm\n",
    "        pass\n",
    "        \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'hmm_7_state' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-4-3f33eb9999ff>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      8\u001b[0m \u001b[0mX\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtranslate_observations_to_indices\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mg1\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'genome1'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      9\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 10\u001b[0;31m \u001b[0mw1\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgenome_pred\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcompute_w_log\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhmm_7_state\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mX\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     11\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mw1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'hmm_7_state' is not defined"
     ]
    }
   ],
   "source": [
    "def getGenomeSequences():\n",
    "    sequences = read_fasta_file(\"/Users/alexiaborchgrevink/Desktop/AU_MachineLearning/Theoretical Excercises/au_ml18/handin3/Handin3_Class1_14/data-handin3/genome1.fa\")\n",
    "    return sequences\n",
    "\n",
    "genome_pred = GenomePrediction()\n",
    "hmm_7_state = genome_pred.init_model()\n",
    "g1 = read_fasta_file('/Users/alexiaborchgrevink/Desktop/AU_MachineLearning/Theoretical Excercises/au_ml18/handin3/Handin3_Class1_14/data-handin3/genome1.fa')\n",
    "g1['genome1'][:]\n",
    "X = translate_observations_to_indices(g1['genome1'])\n",
    "\n",
    "w1 = genome_pred.compute_w_log(hmm_7_state,X)\n",
    "print(w1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
